### PR TITLE
Support auto upgrades for getter only properties

### PR DESCRIPTION
**Pull request by asodja**
We support now automatic upgrades for `@ReplacesEagerProperty` for `Provider<T> getProperty()` like properties.

Fixes https://github.com/gradle/gradle/issues/26540

Tested also with https://github.com/gradle/gradle/pull/29271

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 1]
```diff
@@ -440,7 +467,7 @@ private List<RequestExtra> getJvmRequestExtras(AccessorSpec accessor, Executable
         extras.add(new RequestExtra.OriginatingElement(method));
         extras.add(new RequestExtra.InterceptJvmCalls(interceptorsClassName, BYTECODE_UPGRADE));
         String implementationClass = accessor.generatedClassName;
-        Type newReturnType = TypeUtils.extractReturnType(method);
+        TypeName newPropertyType = TypeName.get(method.getReturnType());
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:470]: 💭 We now pass `TypeName` instead of `Type` to  `PropertyUpgradeClassSourceGenerator.java`, since it contains also type parameters.

Imo we should everywhere get rid of `Type` and use `TypeName`, but that is a bigger change.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 2]
```diff
@@ -225,15 +232,21 @@ private static CodeBlock getGetCall(String propertyGetterName, CallableReturnTyp
...
+                return CodeBlock.of("return new $T<>($N.$N())", SET_PROPERTY_SET_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case MAP_PROPERTY:
-                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case PROPERTY:
                 return CodeBlock.of("return $N.$N().getOrElse($L)", SELF_PARAMETER_NAME, propertyGetterName, TypeUtils.getDefaultValue(returnType.getType()));
+            case PROVIDER: {
+                TypeName newPropertyType = implementationExtra.getNewPropertyType();
+                return newPropertyType instanceof ParameterizedTypeName && isAssignableToFileSystemLocation(((ParameterizedTypeName) newPropertyType).typeArguments.get(0))
+                    ? CodeBlock.of("return $N.$N().map($T::getAsFile).getOrNull()", SELF_PARAMETER_NAME, propertyGetterName, FILE_SYSTEM_LOCATION.asClassName())
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:245]: 💭 We have to specially handle `Provider<RegularFile>` and `Provider<Directory>`. All other cases should be the same.

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy [hunk 3]
```diff
@@ -209,6 +209,58 @@ class PropertyUpgradeCodeGenTest extends InstrumentationCodeGenTest {
...
+        def givenSource = source"""
+            package org.gradle.test;
+
+            import org.gradle.api.provider.*;
+            import org.gradle.api.file.*;
+            import org.gradle.internal.instrumentation.api.annotations.VisitForInstrumentation;
+            import org.gradle.internal.instrumentation.api.annotations.ReplacesEagerProperty;
+            import org.gradle.internal.instrumentation.api.annotations.ReplacedAccessor;
+
+            @VisitForInstrumentation(value = {Task.class})
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy]: ❓ Why is this needed now? Only for the test?

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy [hunk 4]
```diff
@@ -209,6 +209,58 @@ class PropertyUpgradeCodeGenTest extends InstrumentationCodeGenTest {
...
+        assertThat(compilation)
+            .generatedSourceFile(fqName(generatedClass))
+            .containsElementsIn(generatedClass)
+
+        where:
+        upgradedType                    | originalType | getCall                           | imports
+        "Provider<Integer>"             | "Integer"    | "self.getProperty().getOrElse(null)" | []
+        "Provider<RegularFile>"         | "File"       | "self.getProperty().getOrElse(null)" | []
+        "Provider<Directory>"           | "File"       | "self.getProperty().getOrElse(null)" | []
+        "Provider<Map<String, String>>" | "Map"        | "self.getProperty().getOrElse(null)" | []
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy]: 🤔  Oh, `self` is so much of a better name than `thiz`, can we standardize on that instead elsewhere, too?

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 5]
```diff
@@ -364,34 +368,52 @@ private AccessorSpec getAccessorSpec(
...
-                return FILE_COLLECTION.asType();
+                return FILE_COLLECTION.asClassName();
             case DIRECTORY_PROPERTY:
             case REGULAR_FILE_PROPERTY:
-                return Type.getType(File.class);
+                return ClassName.get(File.class);
             case LIST_PROPERTY:
-                return Type.getType(List.class);
+                return ParameterizedTypeName.get(ClassName.get(List.class),
+                    TypeName.get(((DeclaredType) typeMirror).getTypeArguments().get(0))
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:403]: 💅 Let's extract this to some utility method.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java (hunk 1 above)
Comment (lptr) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:470]: That sounds like a good idea. Let's create an issue for that.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 6]
```diff
@@ -225,15 +232,21 @@ private static CodeBlock getGetCall(String propertyGetterName, CallableReturnTyp
...
-                return CodeBlock.of("return new $T<>($N.$N())", SET_PROPERTY_SET_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", SET_PROPERTY_SET_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case MAP_PROPERTY:
-                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asTypeName(), SELF_PARAMETER_NAME, propertyGetterName);
+                return CodeBlock.of("return new $T<>($N.$N())", MAP_PROPERTY_MAP_VIEW.asClassName(), SELF_PARAMETER_NAME, propertyGetterName);
             case PROPERTY:
                 return CodeBlock.of("return $N.$N().getOrElse($L)", SELF_PARAMETER_NAME, propertyGetterName, TypeUtils.getDefaultValue(returnType.getType()));
+            case PROVIDER: {
+                TypeName newPropertyType = implementationExtra.getNewPropertyType();
+                return newPropertyType instanceof ParameterizedTypeName && isAssignableToFileSystemLocation(((ParameterizedTypeName) newPropertyType).typeArguments.get(0))
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:244]: 💅 Let's name this type parameter to make this line more readable. Can we use a utility method to get it?

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 7]
```diff
@@ -216,7 +223,7 @@ private static CodeBlock getDeprecationCodeBlock(PropertyUpgradeRequestExtra req
             .build();
     }
 
-    private static CodeBlock getGetCall(String propertyGetterName, CallableReturnTypeInfo returnType, GradleLazyType upgradedPropertyType) {
+    private static CodeBlock getGetCall(String propertyGetterName, PropertyUpgradeRequestExtra implementationExtra, CallableReturnTypeInfo returnType, GradleLazyType upgradedPropertyType) {
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:226]: 💅 `getGetCall()` sounds a bit weird. What we are doing here is we are generating a getter body. WDYT about `generateGetterBody()` and `generateSetterBody()`?

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy [hunk 8]
```diff
@@ -260,8 +256,8 @@ class PropertyUpgradeCodeGenTest extends InstrumentationCodeGenTest {
         upgradedType                    | originalType | setOriginalType | hasSuppressWarnings | getCall
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy:256]: 💅 WDYT about `getterBody` instead?

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java (hunk 7 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java]: Makes sense, fixed

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy (hunk 8 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy]: Sounds good to me, changed

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java (hunk 6 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java]: Added `TypeUtils. getTypeParameter(TypeName typeName, int index)` that return `Optional<TypeName>`, it's a bit nicer now I think

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle cancel

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java (hunk 5 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java]: Done

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy (hunk 4 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy]: We use it in generated classes. We can also use it in manual interceptors

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy (hunk 3 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCodeGenTest.groovy]: Removed. `@VisitForInstrumentation` is still useful for types that are in Groovy/Java mixed source sets where we can't use annotation processor.

**Comment by asodja** [comment]
@bot-gradle clean

**Review by lptr — APPROVED** [review]
LGTM.
